### PR TITLE
fix: support v5 style type overrides in v6 graphql method

### PR DESCRIPTION
**Pull request by svidgen**
#### Description of changes

Fixes type map for `generateClient` result to include v5 style type overrides. Both of these queries should be correctly typed:

```ts
const listResult = client.graphql<GraphQLQuery<ListThreadsQuery>>({ ... });

// items is correctly typed
const items = result.data?.listThreads;

const subResult = client.graphql<GraphQLSubscription<OnCreateThreadSubscription>>({ ... });
const sub = subResult.subscribe?.({
   // message is correctly typed
   next(message) { ... }
});
```

#### Description of how you validated changes

Added tests.

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/api-graphql/src/types/index.ts [hunk 1]
```diff
@@ -133,23 +142,29 @@ export type GraphQLVariablesV6<
 	? IN
 	: any;
 
+type Sentinel = {
+	sentinel: true;
+};
```
Comment (svidgen) [inline packages/api-graphql/src/types/index.ts:147]: Test cruft. Remove.

File: packages/api-graphql/src/types/index.ts [hunk 2]
```diff
@@ -133,23 +142,29 @@ export type GraphQLVariablesV6<
...
+	FALLBACK_TYPE = unknown,
 	TYPED_GQL_STRING extends string = string
-> = TYPED_GQL_STRING extends GeneratedQuery<any, infer QUERY_OUT>
+> = TYPED_GQL_STRING extends GeneratedQuery<infer IN, infer QUERY_OUT>
 	? Promise<GraphQLResult<QUERY_OUT>>
-	: TYPED_GQL_STRING extends GeneratedMutation<any, infer MUTATION_OUT>
+	: TYPED_GQL_STRING extends GeneratedMutation<infer IN, infer MUTATION_OUT>
 	? Promise<GraphQLResult<MUTATION_OUT>>
-	: TYPED_GQL_STRING extends GeneratedSubscription<any, infer SUB_OUT>
+	: TYPED_GQL_STRING extends GeneratedSubscription<infer IN, infer SUB_OUT>
```
Comment (svidgen) [inline packages/api-graphql/src/types/index.ts:160]: Changed to `any` ➡️ `infer IN` on these lines during debugging, but probably unnecessary.

Look into changing back.

File: packages/api-graphql/__tests__/v6-test.ts [hunk 3]
```diff
@@ -648,10 +875,10 @@ describe('client', () => {
 				.spyOn((raw.GraphQLAPI as any)._api, 'post')
 				.mockImplementation(() => graphqlResponse);
 
-			// Customer would probably not explicitly add `MyType` in their code.
+			// Customer would probably not explicitly add `MyType["result"]` in their code.
 			// But to ensure the test fails if graphql() returns the wrong type, it's explcit here:
-			const result: MyType = client.graphql<MyType>({
-				query: 'query GetWidget { getWidget { name } }',
+			const result: MyType['result'] = client.graphql<MyType>({
+				query: 'query GetWidget($id: ID!) { getWidget(id: $id) { name } }',
```
Comment (svidgen) [inline packages/api-graphql/__tests__/v6-test.ts:881]: Don't recall making this change. Review.

File: packages/api-graphql/src/types/index.ts (hunk 2 above)
Comment (svidgen) [reply packages/api-graphql/src/types/index.ts:160]: Looks like the ignored type inferences work better here. I get some red squigglies when I change them back to `any`. Forgoing the deep-dive as to why and just leaving them as-is for now.

**Review by iartemiev — APPROVED** [review]

**Pull request merged**